src/evaluators/bc/PHAL_Dirichlet_Def.hpp

```diff
@@ -260,39 +260,50 @@ Dirichlet(Teuchos::ParameterList& p) :
 }
 
 // **********************************************************************
-template<typename Traits>
-void Dirichlet<PHAL::AlbanyTraits::DistParamDeriv, Traits>::
-evaluateFields(typename Traits::EvalData dirichletWorkset)
+template <typename Traits>
+void
+Dirichlet<PHAL::AlbanyTraits::DistParamDeriv, Traits>::preEvaluate(
+    typename Traits::EvalData dirichletWorkset)
 {
-  Teuchos::RCP<Thyra_MultiVector> fpV = dirichletWorkset.fpV;
-
   bool trans = dirichletWorkset.transpose_dist_param_deriv;
-  int num_cols = fpV->domain()->dim();
+  if (trans) {
+    auto Vp_bc = dirichletWorkset.Vp_bc;
+    Teuchos::ArrayRCP<Teuchos::ArrayRCP<ST>> Vp_nonconst2dView = Albany::getNonconstLocalData(Vp_bc);
 
-  const std::vector<std::vector<int> >& nsNodes =
-    dirichletWorkset.nodeSets->find(this->nodeSetID)->second;
+    int num_cols = Vp_bc->domain()->dim();
 
-  if (trans) {
-    // For (df/dp)^T*V we zero out corresponding entries in V
-    Teuchos::RCP<Thyra_MultiVector> Vp = dirichletWorkset.Vp_bc;
-    Teuchos::ArrayRCP<Teuchos::ArrayRCP<ST>> Vp_nonconst2dView = Albany::getNonconstLocalData(Vp);
+    const std::vector<std::vector<int> >& nsNodes =
+      dirichletWorkset.nodeSets->find(this->nodeSetID)->second;
 
     for (unsigned int inode = 0; inode < nsNodes.size(); inode++) {
       int lunk = nsNodes[inode][this->offset];
 
       for (int col=0; col<num_cols; ++col) {
-        //(*Vp)[col][lunk] = 0.0;
         Vp_nonconst2dView[col][lunk] = 0.0;
        }
     }
-  } else {
+  }
```

Comment:
Why is there no `else` branch anymore? I don't really understand what `trans` really means, to be fair...

Reply:
I removed it because this step is already done in the pre-evaluate phase for the transpose case.
In the transpose case, the DistParamDeriv evaluation computes  $(\partial_p f)^t \lambda$ whereas $(\partial_p f) q$ is computed in the non-transpose case. Here $f$ is the residual, $p$ is the parameter, $\lambda$ is a vector of the same size as the solution (typically the Lagrange multiplier) and $q$ is a vector having the same size of the distributed parameter $p$.